File: 1639-number-of-ways-to-form-a-target-string-given-a-dictionary/1639-number-of-ways-to-form-a-target-string-given-a-dictionary.py
class Solution:
    MOD = 1000000007

    def numWays(self, words: List[str], target: str) -> int:
        k = len(words[0])
        m = len(target)
        
        freq = [[0] * k for _ in range(26)]
        
        for i in range(k):
            for word in words:
                freq[ord(word[i]) - ord('a')][i] += 1
        
        dp = [[0] * (k + 1) for _ in range(m + 1)]
        dp[0][0] = 1

        for i in range(m + 1):
            for j in range(k + 1):
                if j < k:
                    dp[i][j + 1] = (dp[i][j + 1] + dp[i][j]) % self.MOD
                    if i < m and j < k:
                        dp[i + 1][j + 1] = (dp[i + 1][j + 1] + freq[ord(target[i]) - ord('a')][j] * dp[i][j]) % self.MOD

        return dp[m][k]







# class Solution:
#     def numWays(self, words: List[str], target: str) -> int:
#         MOD = 1000000007
#         m = len(target)
#         k = len(words[0])
#         dp = [[-1] * (k + 1) for _ in range(m + 1)]
#         freq = [[0] * k for _ in range(26)]
        
#         for word in words:
#             for j in range(k):
#                 freq[ord(word[j]) - ord('a')][j] += 1
        
#         def solve(i, j):
#             if i == m:
#                 return 1
#             if j == k:
#                 return 0
#             if dp[i][j] != -1:
#                 return dp[i][j]
#             take = (freq[ord(target[i]) - ord('a')][j] * solve(i + 1, j + 1)) % MOD
#             nottake = solve(i, j + 1) % MOD
#             dp[i][j] = (take + nottake) % MOD
#             return dp[i][j]
        
#         return solve(0, 0)

File: 1639-number-of-ways-to-form-a-target-string-given-a-dictionary/1639-number-of-ways-to-form-a-target-string-given-a-dictionary.java
class Solution {
    public int[][] dp;
    public int m, k;
    public int MOD = 1000000000 + 7;

    public int solve(int i, int j, long[][] freq, String target) {
        if (i == m) return 1;
        if (j == k) return 0;
        if (dp[i][j] != -1) return dp[i][j];

        long take = (freq[target.charAt(i) - 'a'][j] * solve(i+1, j+1, freq, target)) % MOD;
        long nottake = solve(i, j+1, freq, target) % MOD;

        return dp[i][j] = (int)((take + nottake) % MOD);
    }

    public int numWays(String[] words, String target) {
        m = target.length();
        k = words[0].length();
        
        dp = new int[m+1][k+1];
        for (int i = 0; i < m + 1; i++) {
            Arrays.fill(dp[i], -1);
        }

        long[][] freq = new long[26][k];
        for (int i = 0; i < k; i++) {
            for (String x : words) {
                freq[x.charAt(i) - 'a'][i]++;
            }
        }

        return solve(0, 0, freq, target);
    }
}